test/gtest/interpolate.hpp

```diff
@@ -0,0 +1,378 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#include "../driver/tensor_driver.hpp"
+#include "cpu_interpolate.hpp"
+#include "get_handle.hpp"
+#include "random.hpp"
+#include "tensor_holder.hpp"
+#include "verify.hpp"
+#include <cstdint>
+#include <gtest/gtest.h>
+#include <miopen/interpolate.hpp>
+#include <miopen/miopen.h>
+#include <vector>
+
+template <typename T>
+inline std::ostream& operator<<(std::ostream& os, const std::vector<T>& v)
+{
+    os << '{';
+    for(int i = 0; i < v.size(); ++i)
+    {
+        if(i != 0)
+            os << ',';
+        os << v[i];
+    }
+    os << '}';
+    return os;
+}
+
+struct InterpolateTestCase
+{
+    std::vector<size_t> input;
+    std::vector<size_t> size;
+    std::vector<float> scale_factors;
+    miopenInterpolateMode_t mode;
+    bool align_corners;
+
+    friend std::ostream& operator<<(std::ostream& os, const InterpolateTestCase& tc)
+    {
+        return os << " input:" << tc.input << " size:" << tc.size
+                  << " scale_factors:" << tc.scale_factors << " mode:" << tc.mode
+                  << " align_corners:" << tc.align_corners;
+    }
+
+    std::vector<size_t> GetInput() const { return input; }
+};
+
+inline std::vector<InterpolateTestCase> InterpolateTestConfigs()
+{
+    return {
+        {{16, 256, 1, 1, 1}, {32, 32, 32}, {32, 32, 32}, MIOPEN_INTERPOLATE_MODE_TRILINEAR, false},
+        {{16, 256, 1, 1, 1}, {32, 32, 32}, {0, 0, 0}, MIOPEN_INTERPOLATE_MODE_TRILINEAR, true},
+        {{16, 256, 1, 1, 1}, {32, 32, 32}, {0, 0, 0}, MIOPEN_INTERPOLATE_MODE_NEAREST, false},
+        {{16, 256, 1, 1}, {32, 32}, {0, 0}, MIOPEN_INTERPOLATE_MODE_NEAREST, false},
+        {{16, 256, 1, 1}, {32, 32}, {0, 0}, MIOPEN_INTERPOLATE_MODE_BILINEAR, false},
+        {{16, 256, 1, 1}, {32, 32}, {0, 0}, MIOPEN_INTERPOLATE_MODE_BILINEAR, true},
+        {{16, 256, 1, 1}, {32, 32}, {0, 0}, MIOPEN_INTERPOLATE_MODE_BICUBIC, false},
+        {{16, 256, 1, 1}, {32, 32}, {0, 0}, MIOPEN_INTERPOLATE_MODE_BICUBIC, true},
+        {{16, 256, 1}, {32}, {0}, MIOPEN_INTERPOLATE_MODE_NEAREST, false},
+        {{16, 256, 1}, {32}, {0}, MIOPEN_INTERPOLATE_MODE_LINEAR, false},
+        {{16, 256, 1}, {32}, {0}, MIOPEN_INTERPOLATE_MODE_LINEAR, true},
+    };
+}
```

Comment:
I ran gtest and many tests is skipped and failed. I checked source code and the reason for this is that case didn't improve so it didn't pass isApplicable() function. It is okay that you should cover all modes in unit test, but I think it is better if you can only provide test that passed (I don't think AMD reviewers will accept running gtest and failed many tests like that). 
For example with Trilinear mode I saw that forward is not improved so you should only run backward test, you could separate test config for forward and backward like this:
```suggestion
inline std::vector<InterpolateTestCase> InterpolateForwardTestConfigs()
{
    return {
    {{16, 256, 1, 1, 1}, {32, 32, 32}, {32, 32, 32}, MIOPEN_INTERPOLATE_MODE_TRILINEAR, false},
        {{16, 256, 1, 1, 1}, {32, 32, 32}, {0, 0, 0}, MIOPEN_INTERPOLATE_MODE_NEAREST, false},
    };
}
inline std::vector<InterpolateTestCase> InterpolateBackwardTestConfigs()
{
    return {
        {{16, 256, 1, 1, 1}, {32, 32, 32}, {0, 0, 0}, MIOPEN_INTERPOLATE_MODE_NEAREST, false},
    };
}
```

Then in interpolate.cpp, add appropriate test config:
```
INSTANTIATE_TEST_SUITE_P(InterpolateTestSet,
                         InterpolateTestFloat,
                         testing::ValuesIn(InterpolateForwardTestConfigs()));
INSTANTIATE_TEST_SUITE_P(InterpolateTestSet,
                         InterpolateTestFloatBwd,
                         testing::ValuesIn(InterpolateBackwardTestConfigs()));
```

Or with forward, nearest mode, I saw that 4D and 5D tensor will not pass isApplicable(). This test:
``{{16, 256, 1, 1, 1}, {32, 32, 32}, {0, 0, 0}, MIOPEN_INTERPOLATE_MODE_NEAREST, false},`` is 5D tensor and will be skipped. You should change to 3D tensor test for nearest mode.

And please make some changes for other failed/skipped tests.

Reply:
I added the passed tests in commit [368a353](https://github.com/moreh-dev/MIOpen/pull/41/commits/368a353cbba9a6457b851599b6c5bce8312b9ed0). I think I will keep the not isApplicable() cases to show AMD that my constraints are working fine.

Reply:
I think it can be check with driver. Though I think you can modify if AMD requests to change all tests to be passed, so I will mark this as resolved.

Reply:
@hieule88 as all gtest must passed to pass CI/CD when pushing to upstream, you should remove failed test cases

Reply:
removed them